src/ipc/client.rs

```diff
@@ -1,54 +1,106 @@
-use std::env;
-use std::io::{Read, Write};
-use std::net::Shutdown;
-use std::os::unix::net::UnixStream;
-
-use anyhow::{anyhow, bail, Context};
-use niri_ipc::{LogicalOutput, Mode, Output, Reply, Request, Response};
+use anyhow::{bail, Context};
+use niri_ipc::{LogicalOutput, Mode, NiriSocket, Output, Request, Response};
+use serde_json::json;
 
 use crate::cli::Msg;
+use crate::utils::version;
 
 pub fn handle_msg(msg: Msg, json: bool) -> anyhow::Result<()> {
-    let socket_path = env::var_os(niri_ipc::SOCKET_PATH_ENV).with_context(|| {
-        format!(
-            "{} is not set, are you running this within niri?",
-            niri_ipc::SOCKET_PATH_ENV
-        )
-    })?;
+    let mut client = NiriSocket::new()
+        .context("a communication error occured while trying to initialize the socket")?;
 
-    let mut stream =
-        UnixStream::connect(socket_path).context("error connecting to {socket_path}")?;
+    // Default SIGPIPE so that our prints don't panic on stdout closing.
```

Comment:
I intentionally put this after all IPC had been done and before any prints because I'm not sure how this will affect the IPC writes and reads. I suspect it might prevent read/write errors from going through.

Reply:
Hm. How should best to do it then? Handle all IPC logic first, encapsulate the results (including errors) in some variable, then set the signal handler, then do all error printing?

Why do we need it? Is it problematic if the `niri msg` process panics upon stdout hangup? If yes, wouldn't it be better to e.g. lock stdout and actually handle the err case in our code (rather than relying on an unsafe call to libc)?

Reply:
We need it because otherwise things like `niri msg -j focused-window | jq aaaa` (malformed jq command) will spew a Rust backtrace from niri msg in addition to the jq error.

I am not sure what happens to socket IO when it is set though.

Sidenote, there are discussions in Rust to make this more easily settable (e.g. https://github.com/rust-lang/rust/pull/120832), but they are considering CLI apps that don't do socket IO, from my understanding.

Reply:
I think it should be possible in this case to do all the requests before any of the printing, right?

Reply:
It is possible, but somewhat annoying, particularly because of the edge case with version checking. In #294, where i've mainly focused on a more "correct" API (which is easier for e.g. scripting use cases that don't need to handle all possible requests and responses like niri msg does), i've also more clearly separated functionality for sending request, handling version checking, and "outputting" the results of these operations. and then, it's quite trivial to insert a call to ignore `SIGPIPE` at the appropriate point. https://github.com/YaLTeR/niri/blob/f1bfef9a36337696b57efa3e4e41ca04dece1f5b/src/ipc/client.rs#L56-L63